1: Loading the dataset

import numpy as np 

Load the dataset (replace 'dataset.csv' with the actual path to your file) 

Assuming the data columns are in the order: CustomerID, Gender, Age, Annual Income, Spending Score 

data = np.genfromtxt('dataset.csv', delimiter=',', skip_header=1)

You will notice the gender column is all NaNs. Why is that ? How can you solve for it ?


In [1]:
import numpy as np

In [39]:
data = np.genfromtxt("Mall_Customers.csv", dtype=["i8", "U6", "i8", "i8", "i8"], names=["id", "gender", "age", "income", "score"], delimiter=',', skip_header=1)


In [40]:
data

array([(  1, 'Male', 19,  15, 39), (  2, 'Male', 21,  15, 81),
       (  3, 'Female', 20,  16,  6), (  4, 'Female', 23,  16, 77),
       (  5, 'Female', 31,  17, 40), (  6, 'Female', 22,  17, 76),
       (  7, 'Female', 35,  18,  6), (  8, 'Female', 23,  18, 94),
       (  9, 'Male', 64,  19,  3), ( 10, 'Female', 30,  19, 72),
       ( 11, 'Male', 67,  19, 14), ( 12, 'Female', 35,  19, 99),
       ( 13, 'Female', 58,  20, 15), ( 14, 'Female', 24,  20, 77),
       ( 15, 'Male', 37,  20, 13), ( 16, 'Male', 22,  20, 79),
       ( 17, 'Female', 35,  21, 35), ( 18, 'Male', 20,  21, 66),
       ( 19, 'Male', 52,  23, 29), ( 20, 'Female', 35,  23, 98),
       ( 21, 'Male', 35,  24, 35), ( 22, 'Male', 25,  24, 73),
       ( 23, 'Female', 46,  25,  5), ( 24, 'Male', 31,  25, 73),
       ( 25, 'Female', 54,  28, 14), ( 26, 'Male', 29,  28, 82),
       ( 27, 'Female', 45,  28, 32), ( 28, 'Male', 35,  28, 61),
       ( 29, 'Female', 40,  29, 31), ( 30, 'Female', 23,  29, 87),
       ( 31, 'Male', 

2: Understand Customer Demographics

Calculate the average Age, Annual Income, and Spending Score to understand the typical customer profile.

Analyze customer distribution by gender. Does one gender tend to spend more or earn more on average?

Hint: Use NumPy operations like mean(), min(), and max() for this analysis.


In [56]:
age = data["age"]
income = data["income"]
score = data["score"]
gender = data["gender"]

In [49]:
avg_age = np.mean(age)
avg_age

38.85

In [50]:
avg_income = np.mean(income)
avg_income

60.56

In [67]:
avg_score = np.mean(score)
avg_score

50.2

In [62]:
avg_male_income = np.mean(income[gender == "Male"])
avg_male_income

62.22727272727273

In [64]:
avg_female_income = np.mean(income[gender == "Female"])
avg_female_income

59.25

In [65]:
avg_male_score = np.mean(score[gender == "Male"])
avg_male_score

48.51136363636363

In [66]:
avg_female_score = np.mean(score[gender == "Female"])
avg_female_score

51.526785714285715

The average age, annual income and spending score is 38.85, 60.56k and 50.2 respectively.

On average, male's annual income is more than that of female, female's spending score is more than that of male.

3: Identify High-Value Customers

Filter customers with a Spending Score > 80 and calculate their average Annual Income.

Identify the top 10 customers by Spending Score. What do they have in common (e.g., age group, gender)?

Hint: Use boolean indexing and NumPy’s argsort() for ranking.

In [68]:
avg_income_score_80 = np.mean(income[score > 80])
avg_income_score_80

69.06666666666666

In [100]:
sorted = np.argsort(score)
sorted = sorted[::-1]
top_10 = data[sorted][:10]
top_10

array([( 12, 'Female', 35, 19, 99), ( 20, 'Female', 35, 23, 98),
       (186, 'Male', 30, 99, 97), (146, 'Male', 28, 77, 97),
       (168, 'Female', 33, 86, 95), (128, 'Male', 40, 71, 95),
       (  8, 'Female', 23, 18, 94), (142, 'Male', 32, 75, 93),
       (164, 'Female', 31, 81, 93), ( 42, 'Male', 24, 38, 92)],
      dtype=[('id', '<i8'), ('gender', '<U6'), ('age', '<i8'), ('income', '<i8'), ('score', '<i8')])

In [106]:
top_10_gender = gender[sorted][:10]
top_10_age = age[sorted][:10]
top_10_income = income[sorted][:10]

In [97]:
top_10_age_avg = np.mean(top_10_age)
top_10_age_avg

31.1

In [98]:
top_10_income_avg = np.mean(top_10_income)
top_10_income_avg

58.7

In [105]:
male_count = np.sum(top_10_gender == "Male")
male_count

5

The average annual income for customers with a spending score more than 80 is 69.06k.

For customers in top 10 spending score, their average age is 31,1 and average annual income is 58.7k.

For top 10 spending score customers, there are 5 males and 5 females.

4: Explore Relationships Between Features

Compute the pairwise correlations between Age, Annual Income, and Spending Score to uncover key drivers of spending.

Filter young adults (18-25) and calculate their average Spending Score. Compare this with older age groups.

Hint: Correlation can be computed using NumPy’s corrcoef().

In [108]:
corr_age_score = np.corrcoef(age, score)
corr_age_score

array([[ 1.        , -0.32722685],
       [-0.32722685,  1.        ]])

In [109]:
corr_income_score = np.corrcoef(income, score)
corr_income_score

array([[1.        , 0.00990285],
       [0.00990285, 1.        ]])

In [119]:
mask = (age >= 18) & (age <= 25)
young_avg_score = np.mean(score[mask])
young_avg_score

54.94736842105263

In [120]:
old_avg_score = np.mean(score[~mask])
old_avg_score

49.08641975308642

The average score for young adults and older age groups are 54.94 and 49.08 respectively.

5: Customer Segmentation

Group customers into three segments based on their Spending Score:

Low (1-33)

Medium (34-66)

High (67-100)

Compute the average Age and Annual Income for each segment. What are the characteristics of each group?

Propose marketing strategies for each segment.

Hint: Use NumPy’s slicing and aggregation techniques.

In [122]:
low = (score >=1) & (score <= 33)
medium = (score >=34) & (score <= 66)
high = (score >=67) & (score <= 100)

In [124]:
low_avg_age = np.mean(age[low])
low_avg_income = np.mean(income[low])
print(low_avg_age, low_avg_income)

42.87755102040816 67.0


In [125]:
medium_avg_age = np.mean(age[medium])
medium_avg_income = np.mean(income[medium])
print(medium_avg_age, medium_avg_income)

42.01063829787234 53.861702127659576


In [126]:
high_avg_age = np.mean(age[high])
high_avg_income = np.mean(income[high])
print(high_avg_age, high_avg_income)

30.17543859649123 66.0701754385965


Junior tend to have a higher spending score than senior. Annual income does not play an important role in determining spending score.

6: Find Behavioral Similarities

Normalize the dataset (subtract the mean and divide by the standard deviation for each column).

Compute the Euclidean distance between each customer and the customer with the highest Spending Score. Identify customers who exhibit similar behavior.

Hint: Use NumPy’s vectorized operations for distance calculations.

In [140]:
age_norm = (age-np.mean(age))/np.std(age)
income_norm = (income-np.mean(income))/np.std(income)
score_norm = (score-np.mean(score))/np.std(score)

In [141]:
age_norm

array([-1.42456879, -1.28103541, -1.3528021 , -1.13750203, -0.56336851,
       -1.20926872, -0.27630176, -1.13750203,  1.80493225, -0.6351352 ,
        2.02023231, -0.27630176,  1.37433211, -1.06573534, -0.13276838,
       -1.20926872, -0.27630176, -1.3528021 ,  0.94373197, -0.27630176,
       -0.27630176, -0.99396865,  0.51313183, -0.56336851,  1.08726535,
       -0.70690189,  0.44136514, -0.27630176,  0.08253169, -1.13750203,
        1.51786549, -1.28103541,  1.01549866, -1.49633548,  0.7284319 ,
       -1.28103541,  0.22606507, -0.6351352 , -0.20453507, -1.3528021 ,
        1.87669894, -1.06573534,  0.65666521, -0.56336851,  0.7284319 ,
       -1.06573534,  0.80019859, -0.85043527, -0.70690189, -0.56336851,
        0.7284319 , -0.41983513, -0.56336851,  1.4460988 ,  0.80019859,
        0.58489852,  0.87196528,  2.16376569, -0.85043527,  1.01549866,
        2.23553238, -1.42456879,  2.02023231,  1.08726535,  1.73316556,
       -1.49633548,  0.29783176,  2.091999  , -1.42456879, -0.49

In [142]:
income_norm

array([-1.73899919, -1.73899919, -1.70082976, -1.70082976, -1.66266033,
       -1.66266033, -1.62449091, -1.62449091, -1.58632148, -1.58632148,
       -1.58632148, -1.58632148, -1.54815205, -1.54815205, -1.54815205,
       -1.54815205, -1.50998262, -1.50998262, -1.43364376, -1.43364376,
       -1.39547433, -1.39547433, -1.3573049 , -1.3573049 , -1.24279661,
       -1.24279661, -1.24279661, -1.24279661, -1.20462718, -1.20462718,
       -1.16645776, -1.16645776, -1.05194947, -1.05194947, -1.05194947,
       -1.05194947, -1.01378004, -1.01378004, -0.89927175, -0.89927175,
       -0.86110232, -0.86110232, -0.82293289, -0.82293289, -0.82293289,
       -0.82293289, -0.78476346, -0.78476346, -0.78476346, -0.78476346,
       -0.70842461, -0.70842461, -0.67025518, -0.67025518, -0.67025518,
       -0.67025518, -0.63208575, -0.63208575, -0.55574689, -0.55574689,
       -0.55574689, -0.55574689, -0.51757746, -0.51757746, -0.47940803,
       -0.47940803, -0.47940803, -0.47940803, -0.47940803, -0.47

In [143]:
score_norm

array([-0.43480148,  1.19570407, -1.71591298,  1.04041783, -0.39597992,
        1.00159627, -1.71591298,  1.70038436, -1.83237767,  0.84631002,
       -1.4053405 ,  1.89449216, -1.36651894,  1.04041783, -1.44416206,
        1.11806095, -0.59008772,  0.61338066, -0.82301709,  1.8556706 ,
       -0.59008772,  0.88513158, -1.75473454,  0.88513158, -1.4053405 ,
        1.23452563, -0.7065524 ,  0.41927286, -0.74537397,  1.42863343,
       -1.7935561 ,  0.88513158, -1.7935561 ,  1.62274124, -1.4053405 ,
        1.19570407, -1.28887582,  0.88513158, -0.93948177,  0.96277471,
       -0.59008772,  1.62274124, -0.55126616,  0.41927286, -0.86183865,
        0.5745591 ,  0.18634349, -0.12422899, -0.3183368 , -0.3183368 ,
        0.06987881,  0.38045129,  0.14752193,  0.38045129, -0.20187212,
       -0.35715836, -0.00776431, -0.16305055,  0.03105725, -0.16305055,
        0.22516505,  0.18634349,  0.06987881,  0.34162973,  0.03105725,
        0.34162973, -0.00776431, -0.08540743,  0.34162973, -0.12

In [162]:
arr = np.vstack((age_norm, income_norm, score_norm)).T #creating a (200,3) array with normalised age, income and score
euc_dist = np.linalg.norm(arr - arr[score == max(score)], axis=1)
euc_dist

array([2.60142971, 1.23333086, 3.76921616, 1.21828611, 2.30965307,
       1.29364526, 3.61060691, 0.88362925, 4.2686173 , 1.10790218,
       4.02031893, 0.        , 3.65516654, 1.16366029, 3.34195613,
       1.21438449, 2.48575237, 1.67509024, 2.98272515, 0.15753602,
       2.49189884, 1.25310701, 3.7406563 , 1.07408784, 3.58695137,
       0.85964041, 2.72001617, 1.51468853, 2.69134635, 1.05089524,
       4.12274443, 1.48478417, 3.94410976, 1.35936908, 3.49054992,
       1.33542052, 3.2732259 , 1.21464971, 2.91694978, 1.58081869,
       3.36667373, 1.10589185, 2.72670557, 1.68565757, 3.03143718,
       1.71702965, 2.17235359, 2.24663296, 2.3925977 , 2.37097334,
       2.26039964, 1.7560252 , 1.99336143, 2.46944546, 2.52839349,
       2.57890935, 2.4181941 , 3.33136264, 2.20547114, 2.63899874,
       3.18716923, 2.30181951, 3.12177776, 2.3265663 , 2.95560585,
       2.26387309, 2.27452553, 3.27934571, 2.22601803, 2.31232494,
       3.24627439, 2.63618905, 2.89601555, 2.72135538, 2.97096

In [165]:
data[euc_dist <= 1]

array([( 8, 'Female', 23, 18, 94), (12, 'Female', 35, 19, 99),
       (20, 'Female', 35, 23, 98), (26, 'Male', 29, 28, 82)],
      dtype=[('id', '<i8'), ('gender', '<U6'), ('age', '<i8'), ('income', '<i8'), ('score', '<i8')])

Female aged between 23 to 35 with low annual income tend to have a high spending score. 